server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -417,6 +419,22 @@ protected void onCompletion() {
                 }
             }
         }
+
+        @Override
+        public void onCancelled() {
+            try {
+                if (task instanceof CancellableTask t) {
+                    t.ensureNotCancelled();
```

Comment:
Could we just check `t.isCancelled()` rather than throwing and immediately catching an exception?

Reply:
The reason I did this was to catch the exception and add it as a cause in the `DiscardedResponsesException`. In the odd case that the responses are discarded but you are already processing the results getting the `DiscardedResponsesException` without a cause is not very helpful. What do you think?

Reply:
++ yes I realised shortly after writing this that we want to capture the exception for later use, and this is the best way to do that.